feat: [SNOW-3167708] add OIDC workload identity support via service connection (#10)
Snowflake-Labs/snowflake-ado-extension: README.md

Path to the configuration file (`config.toml`) in your repository. The path must be relative to the root of your repository.

### `use-workload-identity`

> Boolean flag to enable OIDC workload identity authentication. When set to `true`, the task will request an OIDC token from Azure DevOps using the specified service connection and configure the Snowflake driver to authenticate via workload identity federation, eliminating the need for storing credentials as secrets. Requires `connected-service-name`. Default is `false`.

Reviewer: Similar to a comment earlier
```suggestion
Boolean flag to enable workload identity federation authentication. When set to `true`, the task will request an OIDC token from Azure DevOps using the specified service connection and configure the Snowflake driver to authenticate with obtained token, eliminating the need for storing credentials as secrets. Requires `connected-service-name`. Default is `false`.
```

Author: Thanks, addressed!

To set up workload identity authentication, follow these steps:

1. **Create an Azure App Registration with a federated credential**:

   Create an App Registration in Microsoft Entra ID and add a federated credential for your ADO service connection. See the [Azure documentation](https://learn.microsoft.com/en-us/entra/workload-id/workload-identity-federation-create-trust) for details.

2. **Create an Azure DevOps service connection**:

   In your ADO project, create an Azure Resource Manager service connection using Workload Identity Federation. Note the service connection name for the pipeline configuration.

3. **Configure Snowflake**:

   Create a service user with OIDC workload identity. The `ISSUER` and `SUBJECT` values come from the OIDC token issued by the service connection. To discover these values, add a debug step to your pipeline:

   ```yaml
   - bash: |
       cat "$SNOWFLAKE_TOKEN_FILE_PATH" | cut -d. -f2 | tr '_-' '/+' | base64 -d 2>/dev/null | python3 -m json.tool
     displayName: 'Debug: inspect OIDC token claims'

>    ```

Reviewer: This feels wrong, I have a gut feeling that there must be a better way to define issuer and subject. Conceptually I think you should first define the issuer and subject on a piece of paper, configure WIF and Snowflake's service user and finally check if everything matches your expectations

Author: Restructured the flow - the issuer (https://vstoken.dev.azure.com/<tenant-id>), subject (sc://<org>/<project>/<connection>), and audience (api://AzureADTokenExchange) all follow documented patterns, so users can now configure Snowflake upfront without running the pipeline first. Moved the token inspection step to a troubleshooting note for when things don't match.

   Then create (or alter) the Snowflake user:

   ```sql
   CREATE USER <username>
     WORKLOAD_IDENTITY = (
       TYPE = OIDC
       ISSUER = '<iss claim from token>'
       SUBJECT = '<sub claim from token>'
       OIDC_AUDIENCE_LIST = ('<aud claim from token>')
     )
     TYPE = SERVICE
     DEFAULT_ROLE = PUBLIC;
   ```

   For more details, see the [Snowflake documentation](https://docs.snowflake.com/en/user-guide/workload-identity-federation).

4. **Configure the pipeline**:

   Add a `config.toml` to your repository (no credentials needed):

   ```toml
   [connections.default]
   account = "<your_account>"

>    user = "<snowflake_service_user>"

Reviewer: can you check if user is even needed? For github OIDC it wasn't

Author: Good call — will verify whether `user` is required for OIDC WIF and update the example accordingly.

Author: Confirmed — `user` is not required. The Snowflake Python connector explicitly includes `WORKLOAD_IDENTITY_AUTHENTICATOR` in the `empty_user_allowed_authenticators` set (connector `connection.py:1748-1753`). Removed `user` from the config example.

Author: it is not needed. Removed

   warehouse = "COMPUTE_WH"

>    role = "SYSADMIN"

Reviewer: ```suggestion
   role = "PUBLIC"
```

Author: addressed. Thanks!

>    role = "SYSADMIN"

Reviewer: ```suggestion
   role = "PUBLIC"
```

Author: addressed
